File: src/lib/contracts.ts
export type AgentProvider = "codex" | "claudeCode" | "openCode";

export type AgentCapability =
  | "usage"
  | "sessions"
  | "liveProcess"
  | "context"
  | "quota"
  | "git"
  | "ports"
  | "mcp"
  | "subagents"
  | "providerMemory"
  | "networkQuota";

export type MeasurementKind =
  | "exact"
  | "estimated"
  | "heuristic"
  | "unavailable";

export type Availability<T> =
  | { state: "available"; value: T; measuredAt: number }
  | {
      state: "stale";
      lastGood: T;
      measuredAt: number;
      staleSince: number;
      reason?: string;
    }
  | {
      state:
        | "unsupported"
        | "notConfigured"
        | "notObserved"
        | "permissionDenied";
      reason?: string;
    };

export type HistoryDomain =
  | "dashboard"
  | "usage"
  | "sessions"
  | "trends"
  | "sources"
  | "quotas"
  | "accounts"
  | "cost"
  | "devices"
  | "settings"
  | "exports"
  | "sync"
  | "alerts";

export type LiveDomain =
  | "host"
  | "liveSessions"
  | "processes"
  | "ports"
  | "mcp"
  | "alerts"
  | "compact";

export type HistoryRevisionEvent = {
  historyRevision: number;
  domains: HistoryDomain[];
};

export type LiveRevisionEvent = {
  liveRevision: number;
  domains: LiveDomain[];
};

export type RuntimeKind = "usage" | "live" | "limits";

export type RuntimeStatus =
  | "notStarted"
  | "starting"
  | "healthy"
  | "degraded"
  | "stopped";

export type RuntimeHealthView = {
  runtime: RuntimeKind;
  status: RuntimeStatus;
  updatedAt: number;
  message?: string;
};

export type ProviderCapabilityView = {
  provider: AgentProvider;
  capabilities: AgentCapability[];
};

export type ReportingTimeZoneView = {
  id: string;
  displayName: string;
  source: "windowsSystem";
};

export type BootstrapView = {
  historyRevision: number;
  liveRevision: number;
  deviceId: string;
  reportingTimeZone: ReportingTimeZoneView;
  implementedCapabilities: ProviderCapabilityView[];
  runtimeHealth: RuntimeHealthView[];
};

export type RefreshScope = "codex";

export type AcceptedRevision = {
  historyRevision: number;
  accepted: boolean;
};

export type Revisioned<T> = {
  asOfHistoryRevision: number;
  data: T;
};

export type LiveRevisioned<T> = {
  asOfLiveRevision: number;
  observedAt: number;
  expiresAt: number;
  data: T;
};

export type DualRevisioned<T> = {
  asOfHistoryRevision: number;
  asOfLiveRevision: number;
  data: T;
};

export type AppError = {
  code: string;
  message: string;
  retryable: boolean;
  sourceId?: string;
};

File: src-tauri/tests/contracts.rs
use tokenometer_lib::domain::{
    AgentProvider, BootstrapView, ProviderCapability, ProviderCapabilityView,
    ReportingTimeZoneSource, ReportingTimeZoneView, RuntimeHealthView, RuntimeKind, RuntimeStatus,
};

#[test]
fn rust_bootstrap_wire_shape_matches_the_shared_frontend_fixture() {
    let actual = BootstrapView {
        history_revision: 7,
        live_revision: 0,
        device_id: "device-synthetic-contract".to_string(),
        reporting_time_zone: ReportingTimeZoneView {
            id: "Synthetic Standard Time".to_string(),
            display_name: "Synthetic Time".to_string(),
            source: ReportingTimeZoneSource::WindowsSystem,
        },
        implemented_capabilities: vec![ProviderCapabilityView {
            provider: AgentProvider::Codex,
            capabilities: vec![
                ProviderCapability::Usage,
                ProviderCapability::Sessions,
                ProviderCapability::Context,
                ProviderCapability::Quota,
            ],
        }],
        runtime_health: vec![
            RuntimeHealthView {
                runtime: RuntimeKind::Usage,
                status: RuntimeStatus::Healthy,
                updated_at: 1_786_147_200_000,
                message: None,
            },
            RuntimeHealthView {
                runtime: RuntimeKind::Live,
                status: RuntimeStatus::NotStarted,
                updated_at: 1_786_147_200_000,
                message: None,
            },
            RuntimeHealthView {
                runtime: RuntimeKind::Limits,
                status: RuntimeStatus::NotStarted,
                updated_at: 1_786_147_200_000,
                message: None,
            },
        ],
    };
    let expected: serde_json::Value =
        serde_json::from_str(include_str!("fixtures/contracts/bootstrap.json")).unwrap();

    assert_eq!(serde_json::to_value(actual).unwrap(), expected);
}
